File: python/completed/euler0042.py
###
#
# Created by Holden on 5/25/2017
#
# COMPLETED on 5/25/2017
#
# Problem 42 - Coded Triangle Numbers
#
###

import math

shift = 64

words = open("../data/x0042.txt", "r")
words_list = []

for line in words:
    words_list.extend(word.strip('"') for word in line.strip().split(","))


def coded_triangle_numbers():
    triangles = 0

    for word in words_list:
        score = sum([(ord(c) - 64) for c in word])
        n = math.floor(math.sqrt(score*2))
        if n*(n+1)/2 == score:
            triangles += 1

    return triangles


print(coded_triangle_numbers())

File: javascript/completed/euler0042.js
import fs from 'fs';

function findWordScore(word, letterVals) {
  let score = 0;
  for (let char of word) {
    if (!letterVals[char]) {
      letterVals[char] = char.charCodeAt(0) - 64;
    }
    score += letterVals[char];
  }
  return score;
}

function isTriangle(num, triangleMap) {
  if (triangleMap[num] === undefined) {
    triangleMap[num] = Number.isInteger((Math.sqrt(8 * num + 1) - 1) / 2);
  }
  return triangleMap[num];
}

export function solve42() {
  const file = fs.readFileSync('./data/x0042.txt', {
    encoding: 'utf-8',
  });
  const words = file.replace(/["]+/g, '').split(',');

  let letterVals = {};
  let triangleMap = {};
  let triangleWords = {};
  let count = 0;

  for (let word of words) {
    const score = findWordScore(word, letterVals);
    if (isTriangle(score, triangleMap)) {
      triangleWords[word] = score;
      count++;
    }
  }

  return count;
}
